File: SDE_sheet/Arrays/15_majority_element_nby2/solution.go
func majorityElement(nums []int) int {
    candidate := 0
    count := 0

    for _, num := range nums {
        if count == 0 {
            candidate = num
            count = 1
        } else if candidate == num {
            count++
        } else {
            count--
        }
    }

    count = 0
    for _, num := range nums {
        if candidate == num {
            count++
        }
    }

    if count > len(nums)/2 {
        return candidate
    }
    return -1
}

File: SDE_sheet/Arrays/15_majority_element_nby2/solution.js
var majorityElement = function(nums) {
    let candidate;
    let count = 0;

    for (const num of nums) {
        if (count === 0) {
            candidate = num;
            count = 1;
        } else if (candidate === num) {
            count++;
        } else {
            count--;
        }
    }

    count = 0;
    for (const num of nums) {
        if (candidate === num) count++;
    }

    return count > nums.length / 2 ? candidate : -1;
};
